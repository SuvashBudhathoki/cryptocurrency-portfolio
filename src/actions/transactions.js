import database from "../firebase/firebase";

//Add_Transaction h

export const addTransaction = (transaction) => ({
  type: "ADD_TRANSACTION",
  transaction,
});

export const startAddTransaction = (transactionData = {}) => {
  return (dispatch) => {
    const {
      currencyName = "",
      units = 0,
      amount = 0,
      createdAt = 0,
    } = transactionData;

    const transaction = {
      currencyName,
      units,
      amount,
      createdAt,
    };

    return database
      .ref("transactions")
      .push(transaction)
      .then((ref) => {
        dispatch(
          addTransaction({
            id: ref.key,
            ...transaction,
          })
        );
      });
  };
};

//Remove_Transaction

export const removeTransaction = ({ id } = {}) => ({
  type: "REMOVE_TRANSACTION",
  id,
});

export const startRemoveTransaction = ({ id }) => {
  return (dispatch) => {
    return database
      .ref(`transactions/${id}`)
      .remove()
      .then(() => {
        dispatch(removeTransaction({ id }));
      });
  };
};

// Edit_Transaction

export const editTransaction = (id, updates) => ({
  type: "EDIT_TRANSACTION",
  id,
  updates,
});

export const startEditTransaction = (id, updates) => {
  return (dispatch) => {
    return database
      .ref(`transactions/${id}`)
      .update(updates)
      .then(() => dispatch(editTransaction(id, updates)));
  };
};

export const setTransactions = (transactions) => ({
  type: "SET_TRANSACTIONS",
  transactions,
});

export const startSetTransactions = () => {
  return (dispatch) => {
    return database
      .ref("transactions")
      .once("value")
      .then((snapshot) => {
        const transactions = [];
        snapshot.forEach((childSnapshot) => {
          transactions.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setTransactions(transactions));
      });
  };
};
