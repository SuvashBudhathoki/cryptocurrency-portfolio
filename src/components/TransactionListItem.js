import React from "react";
import { removeTransaction } from "../actions/transactions";
import { connect } from "react-redux";

const TransactionListItem = ({
  id,
  currencyName,
  amount,
  createdAt,
  units,
  dispatch
}) => (
  <div>
    <h3>{currencyName}</h3>
    <p>
      {units}units ${amount} - {createdAt}
    </p>

    <button
      onClick={() => {
        dispatch(removeTransaction({ id }));
      }}
    >
      Remove
    </button>
  </div>
);

export default connect()(TransactionListItem);
