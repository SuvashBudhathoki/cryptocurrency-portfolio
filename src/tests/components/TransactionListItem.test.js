import React from "react";
import { shallow } from "enzyme";
import TransactionListItem from "../../components/TransactionListItem";
import transactions from "../fixtures/transactions";

//TestCase to check if the individual transaction are displayed

test("should render TransactionListItem correctly", () => {
  const wrapper = shallow(<TransactionListItem {...transactions[0]} />);
  expect(wrapper).toMatchSnapshot();
});
