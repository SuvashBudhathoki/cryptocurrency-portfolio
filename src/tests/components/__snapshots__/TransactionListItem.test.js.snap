// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TransactionListItem correctly 1`] = `
<div>
  <[object Object]
    to="/edit/1"
  >
    <h3>
      Bitcoin
    </h3>
  </[object Object]>
  <p>
    10
    units -
    $3,000.00
    -
    January 1st, 1970
  </p>
</div>
`;
